src/insert_completer.cc

```diff
@@ -481,7 +481,7 @@ bool InsertCompleter::has_candidate_selected() const
 
 void InsertCompleter::try_accept()
 {
-    if (m_completions.is_valid() and m_context.has_client() and has_candidate_selected())
+    if (m_completions.is_valid() and has_candidate_selected())
```

Comment:
> This commit attempts to fix a crash with -f. Specifically when kakoune is run as
> 
> > ./kak -f '2oK.k<c-n><ret><c-n>' /dev/null
> 
> The crash occurs because when <c-n> is pressed for the second time,
> it attempts to use m_completions from the first press of <c-n>.
> This only happens when kakoune is run with -f, because when this is done
> interactively, there is a client, which means that m_completions gets
> reset.  This removes the check that causes that difference.
> 
> I am not *completely* sure that this is the best way to solve the
> problem, since I am not completely sure why that check was put there
> in the first place.

Looks like I added it in 66078243e (Run InsertCompletionHide hook before
insertions that close completion menu, 2022-05-29).
I can tell why I did it this way: I was trying to make the minimal change that would fix the `InsertCompletionHide` hook. That hook only ever needs to run in the client context, so I made sure to only modify those code paths.

So, no particular reason for excluding client-less contexts.
The fix looks great.

I was wondering why the crash does not reproduce interactively (when removing `InsertCompleter::try_accept()`).
This is probably because in interactive mode, completions are quickly updated by the timer (Insert::m_idle_timer), so we don't need to reset them ever.

you could add the test to `test/regression/`

Reply:
I looked at the test infrastructure, but I could not find any way to run to test kakoune with `-f`.

Reply:
I looked at the test infrastructure, but I could not find any way to run to test kakoune with `-f`.

Reply:
> I looked at the test infrastructure, but I could not find any way to run to test kakoune with -f.

I see. We could add a special case to `test/run` but it may not be worth the added complexity.
We generally don't need to test `-f` separately because it's supposed to behave just like interactive mode.
So I think it's fine.